src/storage/src/source/timestamper.rs

```diff
@@ -0,0 +1,325 @@
+                    None => continue,
+                };
+                if bindings.is_empty() {
+                    continue;
+                }
+
+                // Compact as able, relying on all messages being in ascending offset order
+                while bindings.len() > 1
+                    && bindings
+                        .front()
+                        .expect("always at least one binding per partition")
+                        .1
+                        < *offset
+                {
+                    let (_old_timestamp, old_max_offset) = bindings.pop_front().unwrap();
+                    let (_timestamp, incremental_offset) = bindings.front_mut().unwrap();
+                    *incremental_offset += old_max_offset;
+                }
+
+                let (timestamp, max_offset) = bindings
+                    .front()
+                    .expect("always at least one binding per partition");
+                if *offset <= *max_offset {
+                    let old_value =
+                        matched_offsets.insert(partition.clone(), (*timestamp, *offset));
+                    assert_eq!(old_value, None);
+                } else {
+                    assert_eq!(bindings.len(), 1);
+                    // Unable to match any more from this partition
+                    break;
+                }
+            }
+
+            // TODO(#12267): Properly downgrade `since` of the timestamp collection based on the source data collection.
+
+            // XXX: should clamp to round timestamp_frequency??
+            let new_ts = (self.now)();
+            let new_upper = Antichain::from_elem(new_ts + 1);
+
+            let new_bindings: Vec<_> = observed_max_offsets
+                .iter()
+                .filter(|(partition, _offset)| !matched_offsets.contains_key(partition))
+                .map(|(partition, new_max_offset)| {
+                    // Attempt to commit up to max offsets at current timestamp
+                    let current_offset = self
+                        .persisted_timestamp_bindings
+                        .entry(partition.clone())
+                        .or_insert_with(VecDeque::new)
+                        .back()
+                        .map(|(_ts, offset)| *offset)
+                        .unwrap_or_default();
+                    let diff = *new_max_offset - current_offset;
+                    assert!(
+                        diff > 0,
+                        "Diff previously validated to be positive: {:?}",
+                        diff
+                    );
+                    (partition, diff)
+                })
+                .collect();
+
+            if new_bindings.is_empty() {
+                assert_eq!(matched_offsets.len(), observed_max_offsets.len());
+                // This should be a sensible value for the source to downgrade its capability to after emitting the messages that we return
+                let progress = matched_offsets
```

Comment:
Even if the bindings are empty we should still append to the persist shard to bump its upper to `new_upper` and also return `new_upper` to the source to downgrade its capability too.